### PR TITLE
feat(grouped_gemm): add CK work-stealing variant with schedule API

### PR DESCRIPTION
**Pull request by wenchenvincent**
## Summary

Adds a work-stealing (WS) variant of the CK grouped-GEMM kernel (forward +
variable-K backward) for AMD MI355X / MI350. Under all-gather × grouped-GEMM
overlap the static-stride persistent kernel suffers a ~1.7–2.0× slowdown
when RCCL channels contend for the same CUs; with WS enabled, fast CUs
absorb work that the RCCL-throttled CUs would have done, recovering most of
the slowdown at ~2-3% isolated overhead on the 288-shape MoE sweep.

Rebased on top of current `main` (which now includes the Triton WS PR
#353, so the public API is `schedule: str = "static" | "work_steal"`).
This PR wires CK into that same API, so users get WS on either backend by
setting `schedule="work_steal"` on `primus_turbo.ops.grouped_gemm(...)`.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

**Kernel** (`csrc/kernels/grouped_gemm/ck/grouped_gemm_kernel_ws.hpp`):
- Vendored work-stealing variant of `ck_tile::GroupedGemmKernel`. Static-
  stride persistent loop replaced by an `atomicAdd`-based tile claim.
- Two-phase scheme: **phase 1** each CTA drains its XCD's local counter
  for up to `local_per_xcd` claims (preserves L2 locality across XCDs);
  **phase 2** the remainder is claimed via a single global counter
  (cross-XCD work-stealing).
- Counter layout `[xcd0..xcd7, global, done]` — 10 int32 slots.
- Last-out CTA self-resets the counter so the next launch sees a clean
  slate; no host-side `counter.zero_()` needed.
- `FindGroupId` binary search reads precomputed `block_start`/`block_end`
  populated by the existing `compute_grouped_gemm_args` kernel.
- **Defensive bound** — `active_xcds = min(gridDim.x, NUM_XCDS_WS)` used
  for both `xcd_id` and `phase1_total`. Without this cap, launches where
  `gridDim.x < NUM_XCDS_WS` (=8) would leave the tiles in the gap between
  phase 1's actual end and phase 2's start unclaimed → silent wrong
  output. Matches the equivalent fix on the Triton side (PR #402).

**Binding + runner** (`csrc/pytorch/grouped_gemm/ck_grouped_gemm.cpp`,
`csrc/kernels/grouped_gemm/ck/ck_grouped_gemm_kernel_template.h`):
- `ck_grouped_gemm` and `ck_grouped_gemm_variable_k` extended with
  `(work_steal, ws_counter, ws_local_per_xcd)` args.
- `CKGroupedGemmRunner` exposes `run_ws()` alongside `run()`.
- `PRIMUS_TURBO_CHECK` on `ws_local_per_xcd` bounds `[0, INT32_MAX]` to
  reject values that would silently truncate on the `int64 → int32` cast
  passed to the kernel.

**Backend + heuristic**
(`primus_turbo/pytorch/kernels/grouped_gemm/grouped_gemm_impl.py`,
`primus_turbo/pytorch/kernels/grouped_gemm/ws_ck_heuristic.py`):
- `GroupedGEMMCKBackend.can_handle` now advertises
  `schedule in ("static", "work_steal")`; `execute` translates
  `schedule="work_steal"` to the internal `work_steal=True` and picks a
  `ws_local_per_xcd` via the sync-free heuristic.
- Heuristic (`ws_ck_heuristic.py`) resolves the `ws_mode` string to the
  kernel's integer `ws_local_per_xcd`, using `tpc = total_tiles / num_cus`
  as the density signal. Uses tensor metadata only (`a.size(0)`,
  `group_lens.numel()`); no GPU sync.
- `total_m` computed as `a.size(1) if trans_a else a.size(0)` for shape
  correctness under transposed inputs (`can_handle` already rejects
  `trans_a=True` on this backend, but the shape logic is now explicit).
- Wrapper-managed per-device counter cache (`_get_ck_ws_counter`); the
  kernel self-resets so it's a one-shot `torch.zeros` per device.

**Public API** (`primus_turbo/pytorch/ops/grouped_gemm.py`):
- No new kwargs. `schedule="work_steal"` on the existing `grouped_gemm(...)`
  op now dispatches to the CK WS path (in addition to the Triton path
  from PR #353).
- Guard: `schedule="work_steal"` combined with an explicit `num_cu != None`
  raises `ValueError`. WS was designed and characterized for full-device
  launches; partial-grid launches would silently under-utilize the WS
  layout even with the `ACTIVE_XCDS` fix.

**Tests**:
- `tests/pytorch/ops/test_grouped_gemm.py`:
  - `test_grouped_gemm_schedule_work_steal` now parametrized over
    `backend ∈ {TRITON, CK}` — 64 cases per backend asserting forward +
    backward bit-equal vs the static reference.
  - `test_grouped_gemm_schedule_work_steal_single_group` — G=1 smoke on
    both backends (dispatcher special-cases to `gemm_impl`).
  - `test_grouped_gemm_schedule_work_steal_rejects_hipblaslt` — explicit
    HIPBLASLT + `schedule="work_steal"` fails at dispatch (only Triton /
    CK advertise WS via `can_handle`).
  - `test_ck_grouped_gemm_op_ws_num_cu_below_num_xcds` — kernel-level
    regression for the `ACTIVE_XCDS` fix (num_cu ∈ {4, 6, 7, 8, 16}).
- `tests/pytorch/ops/test_ws_ck_heuristic.py` (NEW): 15 pure-Python unit
  tests pinning the CK heuristic's mode-resolution behavior (explicit
  modes, `auto` thresholds for both kernel kinds, degenerate inputs,
  `ValueError` path, `total_tiles` helpers, and the tile / chiplet
  constants).

## Perf (MI355X, 288-shape MoE sweep, isolated kernel, clean methodology)

CK WS isolated overhead vs the static-stride reference (`schedule="static"`):

| Metric | Value |
|---|---:|
| Median | +2.7% |
| p10    | +0.0% |
| p90    | +6.3% |
| Max    | +14.5% |

Under overlap on the canonical bench shape (G=32, M=267424, K=1280, N=2560,
NCCL_MAX_NCHANNELS=default), CK WS recovers >30% of the contended-regime
slowdown vs static.

## Checklist

- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by RuibinCheung** [comment]
Hi @wenchenvincent , Thanks for your contribution. I noticed that code-lint CI failed. You can follow these steps to reformat your code to pass the code-lint CI.
```bash
# make sure you are in the Primus-Turbo root directory
pip install pre-commit
pre-commit install
git commit -m "message"
```

File: primus_turbo/pytorch/ops/grouped_gemm.py [hunk 1]
```diff
@@ -29,6 +29,8 @@ def forward(
         group_offs: torch.Tensor,  # [B+1,] int64
         trans_b: bool,
         num_cu: int | None,
+        work_steal: bool = False,
+        ws_mode: str = "auto",
     ):
         if len(group_lens) == 1:
             assert b.size(0) == 1, f"Expected first dimension to be 1, got {b.size(0)}"
```
[the remark this answers is not in the document]
Comment (wenchenvincent) [reply primus_turbo/pytorch/ops/grouped_gemm.py:35]: The backward path also has the same `if len(group_lens) == 1` short-circuit (lines 69-89 of `primus_turbo/pytorch/ops/grouped_gemm.py`) that dispatches to `gemm_impl` for both `grad_a` and `grad_b` -- `work_steal` and `ws_mode` are ignored in the backward when G=1, identical to the forward. So there's no asymmetry: both forward and backward bypass the WS path for the single-group case. The single-group pytest case (`test_grouped_gemm_work_steal_ck_single_group`) only validates the forward output as you noted, but the underlying dispatch logic is symmetric on both ends.

**Review by xiaobochen-amd — APPROVED** [review]
